### PR TITLE
Discuss the approval of NAP-2 (conda-based packaging for napari)

### PR DESCRIPTION
**Pull request by jaimergp**
#4519 added a [NAP draft](https://napari.org/naps/2-conda-based-packaging.html) to discuss strategies around using `conda` packaging within `napari`, at different levels. We (@goanpeca and I) proposed five milestones to split the discussion:

1. Distributing napari and plugins on conda-forge
2. Building conda-based installers for napari
3. Adding support for conda packages in the plugin manager
4. Enabling in-app napari version updates
5. Deprecating Briefcase-based installers

The discussion already started in the first PR, and some feedback was already incorporated in the text. However, there are many points we still need to discuss before marking this NAP as approved.

This is what I've gathered so far, but please let me know if I should add more talking points and/or action items to the list below.

# Summary of the discussion so far

## Milestone 1 -  Distributing napari and plugins on conda-forge

* We seem to have good agreement on napari being distributed on conda-forge, but we should cover it explicitly in the release guide.
* However, asking plugin developers to submit to conda-forge on their own seems to be too big of a requirement. Better documentation on our end might help, as well as some kind of support in the cookiecutter template, but ultimately this means we can't consider conda-forge the only data source for plugins (at least in the current state, and possibly in the foreseeable future). See milestone 3 for action items.

Action items:

* [ ] Add documentation about conda-forge in the release guide
* [ ] Add documentation about conda-forge in the plugin developer guide
* [ ] Ensure the cookiecutter template has some notion of conda packaging.

## Milestone 2 -  Building conda-based installers for napari

* I think almost everyone agrees that the current `constructor` stack satisfies the installer needs for `napari`
* The multi-environment approach seems to be well received, but we need to ensure that the packaging documentation covers this in a detailed way, since it uses a lot of conda concepts that might be foreign to some devs. This might involve contributing some docs to upstream.
* Creating `napari/packaging` looks like a good idea to keep things tidy and uncoupled

Action items:

* [ ] Add detailed documentation about how the installers are structured and how they work internally
* [ ] Create `napari/packaging` and migrate the conda bundle CI there, along with the relevant metadata 
* [ ] Add PyPI-conda consistency checks to ensure `napari/napari` and `napari/packaging` match

## Milestone 3 -  Adding support for conda packages in the plugin manager

* Supporting `conda` packages exclusively sounds like too big of an ask (see milestone 1)
* Instead, the plugin manager on the bundle should support other plugin sources, like PyPI packages, custom conda channels or local sources.
* Allowing users to install non-conda-forge packages involves some risks we should be prepared for. For example,  via simulated installs to alert users of potential problems, metadata analysis, or recovery mechanisms via `napari-updater`.
* Some comments point out that the plugin manager could become a plugin itself.

_I feel this milestone is one of the most controversial changes, so the discussion here is not over yet._

Action items:

* [ ] Add hybrid support for conda-forge _and_ PyPI packaging. Conda sources are preferred if available; PyPI sources might be marked with warnings or locked unless "advanced mode" is enabled. Some heuristics might anticipate problems or devise a hybrid conda-pip plan where dependencies are satisfied with conda whenever possible first, leaving the rest to pip.
* [ ] Add support for custom conda channels in the UI.
* [ ] Add support for local sources via drag&drop.

## Milestone 4 -  Enabling in-app napari version updates

* If we agree to use multi-environment installations like the infrastructure currently proposes,
  we can implement updates by creating new environments for each version. Multi-env installs would
  allow for different napari "flavors" or "collections" too (not just different versions).
* This is delegated to an external tool (codenamed `napari-updater`) so the logic can be kept
  separate from napari across versions. This tool would also help manage existing installations,
  like reparation tasks ("oh no I installed a plugin and everything broke, what do I do?") and
  cleaning up old installations.
* The development on this tool will happen on `napari/packaging` at first while we prototype
  something quickly, but it will eventually live on its own repo (`napari/updater` or similar) where
  we can make it available for other projects too.

Action items:

- [x] Agree on multi-envs install (e.g. evaluate the balance between flexibility and complexity)
- [ ] Prototype something on `napari/packaging`
- [ ] Refactor prototype into `napari/updater` or equivalent

## Milestone 5 -  Deprecating Briefcase-based installers

* Once we are satisfied with the feature set of the new installers (e.g. pip-install compatibility),
  we can deprecate these pipelines.
* When that's decided, announce in one release, and make it effective for the next one.

Action items:

- [ ] Decide the minimum feature set needed in the conda-based installers so they can replace
      the Briefcase installers. So far, I think it comes down to:
    - [X] Cross-platform compatibility
    - [X] Functional shortcuts in all platforms
    - [ ] Plugin manager support:
        - [X] conda-forge sources
        - [ ] custom conda channels (local or remote, also private)
        - [ ] pypi sources (in a safe way)
        - [ ] local wheels (in a safe way)

### CODE REVIEW

**Comment by tlambert03** [comment]
not quite sure how to review this... The actual changes to the NAP here are just a few lines.

**Review by tlambert03 — CHANGES_REQUESTED** [review]
the nap currently states:

> In the future, we might explore how to deal with PyPI packages within conda in a safe way, but this is an open packaging question that is extremely difficult to tackle robustly, way beyond the scope of this document.

I think that leaves too much room for interpretation for the NAP to be accepted in its current state.  The wording in milestone 3 here is better, but it needs to be brought into the NAP itself before acceptance

**Comment by sofroniewn** [comment]
> Instead, the plugin manager on the bundle should support other plugin sources, like PyPI packages, custom conda channels or local sources.

I think this will be critical to get full napari community support for this NAP

>  Add hybrid support for conda-forge and PyPI packaging. Conda sources are preferred if available; PyPI sources might be marked with warnings or locked unless "advanced mode" is enabled. Some heuristics might anticipate problems or devise a hybrid conda-pip plan where dependencies are satisfied with conda whenever possible first, leaving the rest to pip.

I like this idea. @tlambert03 has called out 

> I think I'd be ok to accept the NAP even if we restrict our support for now to:
>
> - any packages whose dependencies are all on conda-forge, even if the package itself isn't
> - packages that do not themselves require C compilation

I quite like ideas like this the allow us to try and keep the robustness of conda-forge but allow some of the simplicity that comes from PyPI for simple plugins.

I will be out the next 2 weeks so unlikely able to approve this PR myself, but i trust the rest of the community and have already felt like i have been able to give my input in other places and so will be supportive of the outcome if we can approve something 🚀

**Comment by jaimergp** [comment]
I reworded the parts that concern conda+pip, so hopefully that is clearer now!

**Comment by jni** [comment]
@jaimergp I have only had time for a quick look today but I will do my best to do another full pass tomorrow. My impression from the quick look is that the NAP still doesn't address community concerns about installing plugins from a local file or a GitHub repo, and about pip plugins themselves becoming second-class citizens. It should be explicit whether the NAP envisions:

- removing support for these things straight away, or soonish
- removing them only when the community has fully migrated and accepted conda-forge as The One Way (which could be never)
- removing them *only* when we have an easy path for these use cases involving conda
- never removing support, only discouraging it

Whatever the decision is, it should be very clear in the document. Which maybe it is, I haven't had a thorough re-read yet! But if not then that needs updating.

I'll try to have a full re-read tomorrow. Thanks @jaimergp for your efforts and patience! 🙏

**Comment by jni** [comment]
Hi @jaimergp @goanpeca,

I've now gone over the NAP in its entirety. I think this line is still problematic:

> Packages that are only available on PyPI can also be installed as long as:
>  -  The dependencies of the PyPI package are on conda-forge
>  -  The PyPI package is pure Python (no compiled libraries) In the future, we might explore how to deal with PyPI packages within conda in a safer way, but this is an open packaging question that is extremely difficult to tackle robustly.

The community (principally voiced by @tlambert03 and @Czaki, but after further consideration I am also in this camp, see below) really wants an "escape hatch" of sorts to be able to install pip plugins. Therefore, I think in milestone 3 we should *explicitly state* that we will always allow force-installation of pip packages, even if we have to put a big fat warning in front. I don't see the NAP getting accepted without that provision.

What's turned me around is that I am now the proud owner of an M1 Mac, and an increasing proportion of napari users fall in this camp (including @psobolewskiPhD, @perlman, and @alisterburt). And it turns out that to use deep learning on Mac we need tensorflow-metal, which is pip only, or the Apple Silicon version of PyTorch, which is also pip only. Metal PyTorch may *eventually* make it to conda-forge (I know those folks talk to each other), but it might take a very long time (pytorch-cuda took years), but I think the prospects are very slim of tensorflow-metal getting ported to that ecosystem. Given that one of napari's major mission goals is to enable the distribution of state of the art deep learning pipelines to biologists, I don't see how we could go without supporting this library in the bundle.

For this reason, the note:

> At the beginning we will only let pip install pure Python packages, with no dependencies or compiled libraries. Dependencies should be provided from conda-forge whenever possible. Previous experiments on this kind of automations might be useful here [30](https://github.com/napari/napari/blob/c2e33834451a5584bbf13225e759bfaf5627ab31/docs/naps/2-conda-based-packaging.md#user-content-fn-conda-pip-issue-comment-0b9fe0e56a16509ede817b95a376a256) [31](https://github.com/napari/napari/blob/c2e33834451a5584bbf13225e759bfaf5627ab31/docs/naps/2-conda-based-packaging.md#user-content-fn-pamba-0b9fe0e56a16509ede817b95a376a256).

is not sufficient to address concerns about pip packages that are missing in conda-forge. Instead, we should allow "use at your own risk, you might bork the environment" installation of packages with pip. Perhaps `--no-deps` might be a useful default for this setting, forcing users to consider each package and dependency independently, but pure-python-only is too strict.

Additionally, since we are going with a "driver + independent envs" strategy, the risk overall is not that huge because our installer could detect a borked env and refresh the installation, creating a new one where all pip packages are removed.

What do you think about this strategy @jaimergp @goanpeca?

In addition to this, I've pushed a few further changes, some very minor, but a couple of major ones, to (I think) better reflect the concerns voiced during the discussion of this NAP. These are 715da41ed and 440e44d8e .

Regardless of where we land on pip support, all of these concerns need to be discussed in the backward compatibility section, as well as in the alternatives section.

I've changed the status to draft in case we want to merge and continue iterating in a new PR (e.g. should discussion get too long here and we want a fresh start on discussing acceptance). We can go back to "Accepted" after we've landed on a consensus about how to deal with pip.

Thanks for your patience and efforts @jaimergp @goanpeca! I hope we can resolve these final concerns and get this in soon! 🙏

**Comment by jaimergp** [comment]
Thank you so much, @jni! Those are fair points and I am willing to accept your proposed changes. I know we might be erring on the safe side a bit too much, so a warning for the adventurers can be considered enough, given that we'll have a way to resurrect a napari installation easily 🤞

**Comment by tlambert03** [comment]
Strongly agree with @jni there.  Thanks for being open to it @jaimergp.  I do agree that is going too far to state that someone can’t install from pip even after agreeing to some warning that states loudly “doing this may break your environment!”.

The analogy for me is if Apple gave you no way to install an app from outside the App Store, or from unidentified developers.  There would be an uproar 😅

**Comment by jaimergp** [comment]
I have accepted all your suggestions and also added clarification points in other parts of the text that mentioned PyPI. I think the current status should be good for consensus! 🤞

**Comment by jaimergp** [comment]
I have accepted all your suggestions and also added clarification points in other parts of the text that mentioned PyPI. I think the current status should be good for consensus! 🤞 

@tlambert03 @jni

File: docs/naps/2-conda-based-packaging.md
Comment (tlambert03) [inline docs/naps/2-conda-based-packaging.md:105]: quick question: I assume this wording was changed because technically, pip _installs_ packages (it doesn't "have" packages) ... but i'm left wondering if explicitly stating PyPI implies a dropping of support for local `.whl` files.  Since it's currently supported in napari to just drop a wheel file on the plugin window and have it installed, could that be stated here as well?  (again, with the the loud use-at-your-own-risk warnings)

File: docs/naps/2-conda-based-packaging.md
Comment (tlambert03) [inline docs/naps/2-conda-based-packaging.md]:
```suggestion
[^pamba]: https://github.com/tlambert03/pamba
```

put it into a package :)

File: docs/naps/2-conda-based-packaging.md
Comment (tlambert03) [inline docs/naps/2-conda-based-packaging.md:527]: 👍  thank you @jaimergp.  this wording is great and makes me feel much better about all this :)

**Review by tlambert03 — APPROVED** [review]
LGTM for me!  Thanks again for your patience and tolerating all the iteration and requests for some level of (warned) pypi support.

**Review by brisvag — APPROVED** [review]
I like all the changes. I echo @tlambert03 in saying: thanks for the patience! This was a massive job and incorporated a ton of (tricky) feedback. Great work!

File: docs/naps/2-conda-based-packaging.md
Comment (jaimergp) [reply docs/naps/2-conda-based-packaging.md:105]: I am using PyPI as the umbrella term for all package formats that could be hosted there (tarballs, wheels). There's no ill intentions in the change, but thanks for asking! :)

To be clear, we will support tarballs and wheels coming from PyPI and local sources. Custom indices can be part of the conversation as well, if needed.

File: docs/naps/2-conda-based-packaging.md
Comment (jaimergp) [reply docs/naps/2-conda-based-packaging.md]: Oh, nice! Awesome news!

File: docs/naps/2-conda-based-packaging.md
Comment (sofroniewn) [inline docs/naps/2-conda-based-packaging.md:430]: Very excited for this! Once this PR goes in we will create the new repo and grant necessary permissions.

File: docs/naps/2-conda-based-packaging.md
Comment (sofroniewn) [inline docs/naps/2-conda-based-packaging.md:253]: Great - I am very happy with this compromise as our next step

**Review by sofroniewn — APPROVED** [review]
Love this so much! Thanks, I approve and am so excited for this work 🚀

**Review by jni — APPROVED** [review]
❤️ Love it!

**Comment by jni** [comment]
@jaimergp I've merged because I think it's nice to have a PR that *just* flips the approval from draft to approved. And maybe add this PR tot he references come to think of it! 😂 I think that PR can now be merged as soon as it's made. 😊

**Pull request merged**
